### PR TITLE
[SwiftPM] Move the logic for SwiftPM enablement to the platform project

### PR DESCRIPTION
**Pull request by loic-sharma**
In the future, it will be possible for Swift Package Manager to be enabled on one but not all platforms (see https://github.com/flutter/flutter/issues/151567#issuecomment-2455941279).

This moves the `usesSwiftPackageManager` property from the platform-agnostic `Project` to the platform-specific `IosProject` and `MacOSProject`.

This will allow the `IosProject` and `MacOSProject` to return different values for `usesSwiftPackageManager` in the future. For now, both of these projects will always return the same value.

Part of https://github.com/flutter/flutter/issues/151567

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/project_test.dart [hunk 1]
```diff
@@ -376,13 +376,14 @@ void main() {
     });
 
     group('usesSwiftPackageManager', () {
+    // TODO(loic-sharma): Move these to Xcode project unit tests?
```
Comment (loic-sharma) [inline packages/flutter_tools/test/general.shard/project_test.dart:379]: TODO

File: packages/flutter_tools/lib/src/macos/darwin_dependency_management.dart [hunk 2]
```diff
@@ -89,12 +90,8 @@ class DarwinDependencyManagement {
...
 
     final bool useCocoapods;
-    if (_project.usesSwiftPackageManager) {
-      useCocoapods = _usingCocoaPodsPlugin(
-        pluginCount: totalCount,
-        swiftPackageCount: swiftPackageCount,
-        cocoapodCount: podCount,
-      );
+    if (xcodeProject.usesSwiftPackageManager) {
+      useCocoapods = swiftPackageCount < totalCount;
```
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/macos/darwin_dependency_management.dart:94]: TODO: Split into separate PR

File: packages/flutter_tools/lib/src/xcode_project.dart [hunk 3]
```diff
@@ -147,6 +150,40 @@ abstract class XcodeBasedProject extends FlutterProjectPlatform  {
...
+    // https://github.com/flutter/flutter/issues/146957
+    if (parent.isModule) {
+      return false;
+    }
+
+    if (!existsSync()) {
+      return false;
+    }
+
+    // Swift Package Manager requires Xcode 15 or greater.
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:177]: It might be time to require Xcode 15, since Apple has required it for iOS App Store submissions since April https://github.com/flutter/flutter/issues/144582.  See https://github.com/flutter/flutter/pull/146367.



https://developer.apple.com/ios/submit/
> All iOS and iPadOS apps uploaded to App Store Connect must be built with a minimum of Xcode 15 and the iOS 17 SDK.

**Review by jmagman — APPROVED** [review]

**Review by andrewkolos — APPROVED** [review]
LGTM
